rflx/parser/grammar.py

```diff
@@ -429,93 +713,151 @@ def parse_array_aggregate(string: str, location: int, tokens: ParseResults) -> E
+            relation = Greater(left, right, location=locn)
+        elif operator == "in":
+            relation = In(left, right, location=locn)
+        elif operator == "not in":
+            relation = NotIn(left, right, location=locn)
+        else:
+            raise ParseFatalException(string, location, "unexpected relation operator")
+        result.insert(0, relation)
+    return result[0]
 
 
 @fatalexceptions
-def parse_logical_expression(string: str, location: int, tokens: ParseResults) -> Expr:
+def parse_boolean_operator(string: str, location: int, tokens: ParseResults) -> Expr:
     result: List[Expr] = tokens[0]
     while len(result) > 1:
         left = result.pop(0)
         operator = result.pop(0)
         right = result.pop(0)
-        expression: Expr
+
         assert left.location
         assert right.location
+        assert left.location.source == right.location.source
+
         locn = Location(left.location.start, left.location.source, right.location.end)
+
+        bool_expr: Expr
         if operator == "and":
-            expression = And(left, right, location=locn)
+            bool_expr = And(left, right, location=locn)
         elif operator == "or":
-            expression = Or(left, right, location=locn)
+            bool_expr = Or(left, right, location=locn)
         else:
-            raise ParseFatalException(string, location, "unexpected logical operator")
-        result.insert(0, expression)
+            raise ParseFatalException(string, location, "unexpected boolean operator")
+        result.insert(0, bool_expr)
     return result[0]
 
 
-@fatalexceptions
-def parse_mathematical_expression(string: str, location: int, tokens: ParseResults) -> Expr:
+def parse_mathematical_operator(string: str, location: int, tokens: ParseResults) -> Expr:
     result: List[Expr] = tokens[0]
     while len(result) > 1:
         left = result.pop(0)
         operator = result.pop(0)
         right = result.pop(0)
-        assert left.location, f'expression "{left}" without location'
-        assert right.location, f'expression "{right}" without location'
-        assert left.location.source == right.location.source, "expression with different source"
+
+        assert left.location
+        assert right.location
+        assert left.location.source == right.location.source
+
         locn = Location(left.location.start, left.location.source, left.location.end)
-        expression: Expr
+
+        math_expr: Expr
         if operator == "+":
-            expression = Add(left, right)
-            expression.location = locn
+            math_expr = Add(left, right)
+            math_expr.location = locn
         elif operator == "-":
-            expression = Sub(left, right, locn)
+            math_expr = Sub(left, right, locn)
         elif operator == "*":
-            expression = Mul(left, right)
-            expression.location = locn
+            math_expr = Mul(left, right)
+            math_expr.location = locn
```

Comment:
Is there a reason these expressions do not have a location in their constructor?